Mathlib/Probability/Kernel/RadonNikodym.lean

```diff
@@ -0,0 +1,443 @@
+      - Real.toNNReal (1 - rnDerivAux κ (κ + η) a x) * rnDeriv κ η a x) := by
+  change Measurable ((Function.uncurry fun a b ↦
+    ENNReal.ofReal (rnDerivAux κ (κ + η) a b)
+    - ENNReal.ofReal (1 - rnDerivAux κ (κ + η) a b) * rnDeriv κ η a b) ∘ (fun b ↦ (a, b)))
+  exact (measurable_singularPart_fun κ η).comp measurable_prod_mk_left
+
+lemma singularPart_compl_mutuallySingularSetSlice (κ η : kernel α γ) [IsSFiniteKernel κ]
+    [IsSFiniteKernel η] (a : α) :
+    singularPart κ η a (mutuallySingularSetSlice κ η a)ᶜ = 0 := by
+  have h_coe : ∀ b, (Real.toNNReal b : ℝ≥0∞) = ENNReal.ofReal b := fun _ ↦ rfl
+  rw [singularPart, kernel.withDensity_apply', lintegral_eq_zero_iff, EventuallyEq,
+    ae_restrict_iff]
+  all_goals simp_rw [h_coe]
+  rotate_left
+  · exact measurableSet_preimage (measurable_singularPart_fun_right κ η a)
+      (measurableSet_singleton _)
+  · exact measurable_singularPart_fun_right κ η a
+  · exact measurable_singularPart_fun κ η
+  refine ae_of_all _ (fun x hx ↦ ?_)
+  simp only [mem_compl_iff, mem_setOf_eq] at hx
+  simp_rw [rnDeriv]
+  rw [← ENNReal.ofReal_div_of_pos, div_eq_inv_mul, ← ENNReal.ofReal_mul, ← mul_assoc,
+    mul_inv_cancel, one_mul, tsub_self]
+  · rfl
+  · rw [ne_eq, sub_eq_zero]
+    exact Ne.symm hx
+  · simp [rnDerivAux_le_one (le_add_of_nonneg_right bot_le)]
+  · simp [lt_of_le_of_ne (rnDerivAux_le_one (le_add_of_nonneg_right bot_le)) hx]
+
+lemma singularPart_subset_compl_mutuallySingularSetSlice [IsFiniteKernel κ]
+    [IsFiniteKernel η] {a : α} {s : Set γ} (hs : s ⊆ (mutuallySingularSetSlice κ η a)ᶜ) :
+    singularPart κ η a s = 0 := by
+  exact measure_mono_null hs (singularPart_compl_mutuallySingularSetSlice κ η a)
+
+lemma singularPart_subset_mutuallySingularSetSlice [IsFiniteKernel κ]
+    [IsFiniteKernel η] {a : α} {s : Set γ} (hsm : MeasurableSet s)
+    (hs : s ⊆ mutuallySingularSetSlice κ η a) :
+    singularPart κ η a s = κ a s := by
+  have hs' : ∀ x ∈ s, rnDerivAux κ (κ + η) a x = 1 := fun _ hx ↦ hs hx
+  rw [singularPart, kernel.withDensity_apply']
+  swap; · exact measurable_singularPart_fun κ η
+  calc
+    ∫⁻ x in s, ↑(Real.toNNReal (rnDerivAux κ (κ + η) a x)) -
+      ↑(Real.toNNReal (1 - rnDerivAux κ (κ + η) a x)) * rnDeriv κ η a x
+      ∂(κ + η) a
+    = ∫⁻ _ in s, 1 ∂(κ + η) a := by
+        refine set_lintegral_congr_fun hsm (ae_of_all _ fun x hx ↦ ?_)
+        simp [hs' x hx]
+  _ = (κ + η) a s := by simp
+  _ = κ a s := by
+        suffices η a s = 0 by simp [this]
+        exact measure_mono_null hs (measure_mutuallySingularSetSlice κ η a)
+
+lemma withDensity_rnDeriv_mutuallySingularSetSlice (κ η : kernel α γ) [IsFiniteKernel κ]
+    [IsFiniteKernel η] (a : α) :
+    withDensity η (rnDeriv κ η) a (mutuallySingularSetSlice κ η a) = 0 := by
+  rw [kernel.withDensity_apply']
+  · exact set_lintegral_measure_zero _ _ (measure_mutuallySingularSetSlice κ η a)
+  · exact measurable_rnDeriv κ η
+
+lemma withDensity_rnDeriv_subset_mutuallySingularSetSlice [IsFiniteKernel κ]
```

Comment:
```suggestion
lemma withDensity_rnDeriv_of_subset_mutuallySingularSetSlice [IsFiniteKernel κ]
```
Otherwise, it sounds as iff `withDensity_rnDeriv` was a subset of something (which doesn't make sense, of course). Same thing in the next lemma.